File: generate-sitemap.js
#!/usr/bin/env node

/**
 * Automatic Sitemap Generator for Chang Diving Center
 * Scans all HTML files and generates XML sitemap with proper URLs
 * Excludes noindex pages and redirect pages
 */

const fs = require("fs");
const path = require("path");

// Configuration
const config = {
  baseUrl: "https://changdiving.com",
  languages: ["en", "de", "th"],
  outputFile: "sitemap.xml",
  excludePatterns: [
    "node_modules",
    ".git",
    "docs",
    "img",
    "js",
    "fonts",
    "style.css",
    "robots.txt",
    "_redirects.txt",
  ],
};

// Priority mapping for different page types
const priorityMap = {
  "index.html": 1.0, // Homepage
  about: 0.8, // About pages
  courses: 0.9, // Course pages
  "day-trips": 0.8, // Day trip pages
  equipment: 0.7, // Equipment pages
  "dive-sites": 0.7, // Dive sites
  posts: 0.6, // Blog posts
  faqs: 0.5, // FAQ pages
  contact: 0.7, // Contact
  prices: 0.8, // Pricing
  default: 0.5, // Everything else
};

// Change frequency for different page types
const changefreqMap = {
  "index.html": "weekly",
  about: "monthly",
  courses: "monthly",
  "day-trips": "monthly",
  equipment: "monthly",
  "dive-sites": "monthly",
  posts: "weekly",
  faqs: "monthly",
  contact: "monthly",
  prices: "weekly",
  default: "monthly",
};

function shouldExclude(filePath) {
  return config.excludePatterns.some((pattern) => filePath.includes(pattern));
}

function getPriority(urlPath) {
  // Homepage-Check zuerst - höchste Priorität
  if (urlPath === "" || urlPath === "/" || urlPath.endsWith("/index.html")) {
    return 1.0;
  }

  // Language homepage check
  if (urlPath.match(/^\/(en|de|th)\/$/)) {
    return 1.0;
  }

  // Dann andere Checks...
  for (const [key, priority] of Object.entries(priorityMap)) {
    if (key === "default") continue;
    if (urlPath.includes(key)) {
      return priority;
    }
  }
  return priorityMap.default;
}

function getChangeFreq(urlPath) {
  // Homepage-Check zuerst - häufiger Updates
  if (urlPath === "" || urlPath === "/" || urlPath.endsWith("/index.html")) {
    return "weekly";
  }

  // Language homepage check
  if (urlPath.match(/^\/(en|de|th)\/$/)) {
    return "weekly";
  }

  // Dann andere Checks...
  for (const [key, freq] of Object.entries(changefreqMap)) {
    if (key === "default") continue;
    if (urlPath.includes(key)) {
      return freq;
    }
  }
  return changefreqMap.default;
}

/**
 * Check if HTML file should be excluded from sitemap
 * @param {string} filePath - Path to the HTML file
 * @returns {boolean} - True if should be excluded
 */
function shouldExcludeHtmlFile(filePath) {
  try {
    const content = fs.readFileSync(filePath, "utf8");

    // Check for noindex meta tag
    const noindexRegex =
      /<meta[^>]*name=["']robots["'][^>]*content=["'][^"']*noindex[^"']*["'][^>]*>/i;
    if (noindexRegex.test(content)) {
      console.log(`  ⚠️  Excluding (noindex): ${filePath}`);
      return true;
    }

    // Check for JavaScript redirects (but exclude mailto: links)
    const redirectRegex =
      /window\.location\.(replace|href)\s*=\s*["'](?!mailto:)(https?:|\.|\/)/i;
    if (redirectRegex.test(content)) {
      console.log(`  🔄 Excluding (redirect): ${filePath}`);
      return true;
    }
    // Also check for location.replace with http/https (but not mailto)
    const locationReplaceRegex =
      /location\.replace\s*\(\s*["'](?!mailto:)(https?:|\.|\/)/i;
    if (locationReplaceRegex.test(content)) {
      console.log(`  🔄 Excluding (redirect): ${filePath}`);
      return true;
    }

    // Check for 404 pages (by path or title)
    // Only exclude if it's actually in a 404 directory or has explicit 404 title
    if (filePath.includes("/404/") || filePath.includes("/410/")) {
      console.log(`  🚫 Excluding (404/410 page): ${filePath}`);
      return true;
    }
    
    // Check for explicit 404 page title (not just "not found" text in content)
    const title404Regex = /<title[^>]*>.*404.*<\/title>/i;
    if (title404Regex.test(content)) {
      console.log(`  🚫 Excluding (404 in title): ${filePath}`);
      return true;
    }

    return false;
  } catch (error) {
    console.log(`  ❌ Error reading ${filePath}: ${error.message}`);
    return true; // Exclude if we can't read it
  }
}

function scanDirectory(dir, urls = []) {
  const files = fs.readdirSync(dir);

  for (const file of files) {
    const filePath = path.join(dir, file);
    const stat = fs.statSync(filePath);

    if (shouldExclude(filePath)) {
      continue;
    }

    if (stat.isDirectory()) {
      scanDirectory(filePath, urls);
    } else if (file === "index.html") {
      // Check if this HTML file should be excluded
      if (shouldExcludeHtmlFile(filePath)) {
        continue;
      }

      // Convert file path to URL
      let urlPath = filePath.replace(/\\/g, "/");

      // Remove index.html from path
      urlPath = urlPath.replace("/index.html", "");

      // Add leading slash if not present
      if (!urlPath.startsWith("/")) {
        urlPath = "/" + urlPath;
      }

      // Handle root index.html and language homepages
      if (urlPath === "/") {
        urlPath = "";
      } else {
        // Add trailing slash for consistency with actual website URLs
        if (!urlPath.endsWith("/")) {
          urlPath += "/";
        }
      }

      // Special handling for language homepages
      const langHomepageMatch = urlPath.match(/^\/(en|de|th)\/$/);
      if (langHomepageMatch) {
        // This is a language homepage - should have highest priority
        urlPath = "/" + langHomepageMatch[1] + "/";
      }

      const fullUrl = config.baseUrl + urlPath;
      const priority = getPriority(urlPath);
      const changefreq = getChangeFreq(urlPath);

      // Always use file modification time for accurate lastmod dates
      const lastmod = stat.mtime.toISOString().split("T")[0];

      urls.push({
        url: fullUrl,
        lastmod,
        changefreq,
        priority,
      });

      // Debug output for homepages
      if (urlPath.match(/^\/(en|de|th)\/$/)) {
        console.log(`  🏠 Homepage: ${fullUrl} - Lastmod: ${lastmod}`);
      }
    }
  }

  return urls;
}

function generateSitemap(urls) {
  let xml = `<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
`;

  // Sort URLs by priority (highest first), then alphabetically
  urls.sort((a, b) => {
    if (a.priority !== b.priority) {
      return b.priority - a.priority;
    }
    return a.url.localeCompare(b.url);
  });

  for (const urlData of urls) {
    xml += `  <url>
    <loc>${urlData.url}</loc>
    <lastmod>${urlData.lastmod}</lastmod>
    <changefreq>${urlData.changefreq}</changefreq>
    <priority>${urlData.priority}</priority>
  </url>
`;
  }

  xml += "</urlset>";
  return xml;
}

function generateLanguageStats(urls) {
  const stats = {};

  for (const lang of config.languages) {
    stats[lang] = urls.filter((u) => u.url.includes(`/${lang}/`)).length;
  }

  // Count homepage and language-neutral pages
  stats["homepage"] = urls.filter((u) => u.url === config.baseUrl).length;
  stats["total"] = urls.length;

  return stats;
}

// Main execution
console.log("🗺️  Chang Diving Center Sitemap Generator (Enhanced)");
console.log("==================================================");

try {
  console.log("📁 Scanning directories for HTML files...");
  console.log("🔍 Checking for noindex and redirect pages...");
  const urls = scanDirectory(".");

  console.log("\n📊 Statistics:");
  const stats = generateLanguageStats(urls);

  for (const [lang, count] of Object.entries(stats)) {
    console.log(`   ${lang.toUpperCase()}: ${count} pages`);
  }

  console.log("\n🏆 Top priority pages:");
  const topPages = urls
    .filter((u) => u.priority >= 0.8)
    .sort((a, b) => b.priority - a.priority)
    .slice(0, 10);

  topPages.forEach((page) => {
    console.log(`   ${page.priority} - ${page.url}`);
  });

  console.log("\n📝 Generating XML sitemap...");
  const sitemapXml = generateSitemap(urls);

  fs.writeFileSync(config.outputFile, sitemapXml);
  console.log(`✅ Sitemap generated: ${config.outputFile}`);
  console.log(`🌐 Total URLs: ${urls.length}`);
  console.log(`📍 Base URL: ${config.baseUrl}`);

  // Show file size
  const stats_file = fs.statSync(config.outputFile);
  const fileSize = (stats_file.size / 1024).toFixed(2);
  console.log(`📊 File size: ${fileSize} KB`);

  console.log("\n✨ Enhanced features:");
  console.log("   • Excluded noindex pages");
  console.log("   • Excluded redirect pages");
  console.log("   • Excluded 404 pages");
} catch (error) {
  console.error("❌ Error generating sitemap:", error);
  process.exit(1);
}

File: generate_sitemap.py
#!/usr/bin/env python3
"""
Generate sitemap.xml with accurate lastmod dates from Git history.
This script reads the last commit date for each HTML file and generates
a sitemap with accurate modification dates.

Excludes: 404/410 error pages, /search/ utility pages (SEO: do not index).
Keeps: search-recovery, about-search-recovery (content pages).

Usage:
    python3 generate_sitemap.py
"""

import os
import subprocess
from datetime import datetime
from pathlib import Path
import xml.etree.ElementTree as ET
from xml.dom import minidom

# Base URL
BASE_URL = "https://changdiving.com"

# Directories to scan for HTML files
INCLUDE_DIRS = ["en", "de", "th"]

# Files/directories to exclude
EXCLUDE_PATTERNS = [
    ".git",
    "node_modules",
    ".cursor",
    "terminals",
    "docs",
    "fonts",
    "img",
    "js",
    "css",
    "/404/",   # Error pages - do not index (SEO)
    "/410/",   # Gone pages - do not index (SEO)
    "/search/",  # Search utility pages - noindex (not /search-recovery/)
]

# Priority and changefreq rules
PRIORITY_RULES = {
    "index.html": {"priority": "1.0", "changefreq": "weekly"},
    "/en/": {"priority": "1.0", "changefreq": "weekly"},
    "/de/": {"priority": "1.0", "changefreq": "weekly"},
    "/th/": {"priority": "1.0", "changefreq": "weekly"},
    "/courses/": {"priority": "0.9", "changefreq": "monthly"},
    "/day-trips/": {"priority": "0.9", "changefreq": "monthly"},
    "/dive-sites/": {"priority": "0.8", "changefreq": "monthly"},
    "/posts/": {"priority": "0.7", "changefreq": "monthly"},
}


def get_git_last_modified(file_path):
    """
    Get the last modification date of a file from Git history.
    Returns ISO date string (YYYY-MM-DD).
    """
    try:
        result = subprocess.run(
            ["git", "log", "-1", "--format=%cI", "--", file_path],
            capture_output=True,
            text=True,
            check=True,
        )
        date_str = result.stdout.strip()
        if date_str:
            # Parse ISO date and return YYYY-MM-DD
            dt = datetime.fromisoformat(date_str.replace("Z", "+00:00"))
            return dt.strftime("%Y-%m-%d")
    except (subprocess.CalledProcessError, ValueError) as e:
        print(f"Warning: Could not get Git date for {file_path}: {e}")
    
    # Fallback to file system modification time
    try:
        mtime = os.path.getmtime(file_path)
        return datetime.fromtimestamp(mtime).strftime("%Y-%m-%d")
    except:
        return datetime.now().strftime("%Y-%m-%d")


def get_priority_and_changefreq(url_path):
    """
    Determine priority and changefreq based on URL path.
    """
    for pattern, rules in PRIORITY_RULES.items():
        if pattern in url_path:
            return rules
    
    # Default values
    return {"priority": "0.6", "changefreq": "monthly"}


def should_exclude(path):
    """
    Check if a path should be excluded.
    """
    path_str = str(path).replace("\\", "/")
    for pattern in EXCLUDE_PATTERNS:
        if pattern in path_str:
            return True
    return False


def find_html_files():
    """
    Find all HTML files in the repository.
    Returns a list of (file_path, url_path) tuples.
    """
    html_files = []
    
    # Add main index.html
    if os.path.exists("index.html"):
        html_files.append(("index.html", "/"))
    
    # Scan language directories
    for lang_dir in INCLUDE_DIRS:
        if not os.path.exists(lang_dir):
            continue
        
        for root, dirs, files in os.walk(lang_dir):
            # Remove excluded directories from search
            dirs[:] = [d for d in dirs if not should_exclude(os.path.join(root, d))]
            
            for file in files:
                if file.endswith(".html"):
                    file_path = os.path.join(root, file)
                    
                    # Skip if excluded
                    if should_exclude(file_path):
                        continue
                    
                    # Convert file path to URL path
                    url_path = "/" + file_path.replace("\\", "/")
                    if file == "index.html":
                        url_path = url_path.replace("/index.html", "/")
                    
                    html_files.append((file_path, url_path))
    
    return html_files


def generate_sitemap():
    """
    Generate sitemap.xml with accurate Git modification dates.
    """
    print("🔍 Finding HTML files...")
    html_files = find_html_files()
    print(f"   Found {len(html_files)} HTML files")
    
    print("\n📅 Getting Git modification dates...")
    
    # Create XML structure
    urlset = ET.Element("urlset")
    urlset.set("xmlns", "http://www.sitemaps.org/schemas/sitemap/0.9")
    
    # Add URLs
    for file_path, url_path in sorted(html_files, key=lambda x: x[1]):
        # Get last modified date from Git
        lastmod = get_git_last_modified(file_path)
        
        # Get priority and changefreq
        rules = get_priority_and_changefreq(url_path)
        
        # Create URL element
        url = ET.SubElement(urlset, "url")
        
        loc = ET.SubElement(url, "loc")
        loc.text = BASE_URL + url_path
        
        lastmod_elem = ET.SubElement(url, "lastmod")
        lastmod_elem.text = lastmod
        
        changefreq = ET.SubElement(url, "changefreq")
        changefreq.text = rules["changefreq"]
        
        priority = ET.SubElement(url, "priority")
        priority.text = rules["priority"]
        
        print(f"   ✓ {url_path:<50} {lastmod}")
    
    # Pretty print XML
    xml_string = ET.tostring(urlset, encoding="unicode")
    dom = minidom.parseString(xml_string)
    pretty_xml = dom.toprettyxml(indent="  ")
    
    # Remove extra blank lines
    pretty_xml = "\n".join([line for line in pretty_xml.split("\n") if line.strip()])
    
    # Write to file
    with open("sitemap.xml", "w", encoding="utf-8") as f:
        f.write(pretty_xml)
    
    print(f"\n✅ Generated sitemap.xml with {len(html_files)} URLs")
    print(f"   File: {os.path.abspath('sitemap.xml')}")


def main():
    """
    Main function.
    """
    print("=" * 70)
    print("🗺️  SITEMAP GENERATOR WITH GIT DATES")
    print("=" * 70)
    
    # Check if we're in a git repository
    if not os.path.exists(".git"):
        print("❌ Error: Not in a Git repository!")
        return 1
    
    try:
        generate_sitemap()
        print("\n✨ Done!")
        return 0
    except Exception as e:
        print(f"\n❌ Error: {e}")
        import traceback
        traceback.print_exc()
        return 1


if __name__ == "__main__":
    exit(main())
